br/pkg/task/stream.go

```diff
@@ -0,0 +1,456 @@
+	defer func() {
+		if err != nil {
+			mgr.Close()
+		}
+	}()
+
+	client, err := backup.NewBackupClient(ctx, mgr)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	backend, err := storage.ParseBackend(cfg.Storage, &cfg.BackendOptions)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	opts := storage.ExternalStorageOptions{
+		NoCredentials:   cfg.NoCreds,
+		SendCredentials: cfg.SendCreds,
+		SkipCheckPath:   cfg.SkipCheckPath,
+	}
+	if err = client.SetStorage(ctx, backend, &opts); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s := &streamStartMgr{
+		Cfg: cfg,
+		mgr: mgr,
+		bc:  client,
+	}
+	return s, nil
+}
+
+func (s *streamStartMgr) close() {
+	s.mgr.Close()
+}
+
+func (s *streamStartMgr) setLock(ctx context.Context) error {
+	return s.bc.SetLockFile(ctx)
+}
+
+// checkStartTS checks that startTS should be smaller than currentTS,
+// and endTS is larger than currentTS.
+func (s *streamStartMgr) checkStartTS(ctx context.Context) error {
+	currentTS, err := s.getTS(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if currentTS <= s.Cfg.startTS || s.Cfg.endTS <= currentTS {
+		return errors.Annotatef(berrors.ErrInvalidArgument,
+			"invalid timestamps, startTS %llu should be smaller than currentTS %llu",
+			s.Cfg.startTS, currentTS)
+	}
+	if s.Cfg.endTS <= currentTS {
+		return errors.Annotatef(berrors.ErrInvalidArgument,
+			"invalid timestamps, endTS %llu should be larger than currentTS %llu",
+			s.Cfg.endTS, currentTS)
+	}
+
+	return nil
+}
+
+// setGCSafePoint specifies currentTS should belong to (gcSafePoint, currentTS),
+// and set startTS as a serverSafePoint to PD
+func (s *streamStartMgr) setGCSafePoint(ctx context.Context) error {
+	if err := s.checkStartTS(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	err := utils.CheckGCSafePoint(ctx, s.mgr.GetPDClient(), s.Cfg.startTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	sp := utils.BRServiceSafePoint{
+		ID:       utils.MakeSafePointID(),
+		TTL:      s.Cfg.safePointTTL,
+		BackupTS: s.Cfg.startTS,
+	}
+	err = utils.UpdateServiceSafePoint(ctx, s.mgr.GetPDClient(), sp)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("set stream safePoint", zap.Object("safePoint", sp))
+	return nil
+}
+
+func (s *streamStartMgr) getTS(ctx context.Context) (uint64, error) {
+	p, l, err := s.mgr.PdController.GetPDClient().GetTS(ctx)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	return oracle.ComposeTS(p, l), nil
+}
+
+func (s *streamStartMgr) buildObserveRanges(ctx context.Context) ([]kv.KeyRange, error) {
+	dRanges, err := stream.BuildObserveDataRanges(
+		s.mgr.GetStorage(),
+		s.Cfg.TableFilter,
+		s.Cfg.startTS,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	mRange, err := stream.BuildObserveMetaRange()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rs := append([]kv.KeyRange{*mRange}, dRanges...)
+	sort.Slice(rs, func(i, j int) bool {
+		return bytes.Compare(rs[i].StartKey, rs[j].EndKey) < 0
```

Comment:
```suggestion
		return bytes.Compare(rs[i].StartKey, rs[j].StartKey) < 0
```

Indeed this works (because there can rarely be overlapping ranges), but if there were, it can cause `!less(x, y) != less(y, x)`, which (IMO) is terrible😱. If there isn't overlapping ranges, sort them just by `StartKey` is fine.